docs/architecture/adr-004-split-denomination-keys.md

```diff
@@ -0,0 +1,73 @@
+# ADR 004: Split Denomination Keys
+
+## Changelog
+
+- 2020-01-08: Initial version
+- 2020-01-09: Alterations to handle vesting accounts
+- 2020-01-14: Updates from review feedback
+
+## Context
+
+With permissionless IBC, anyone will be able to send arbitrary denominations to any other account. Currently, all non-zero balances are stored along with the account in an `sdk.Coins` struct, which creates a potential denial-of-service concern, as too many denominations will become expensive to load & store each time the account is modified. See issues [5467](https://github.com/cosmos/cosmos-sdk/issues/5467) and [4982](https://github.com/cosmos/cosmos-sdk/issues/4982) for additional context.
+
+Simply rejecting incoming deposits after a denomination count limit doesn't work, since it opens up a griefing vector: someone could send a user lots of nonsensical coins over IBC, and then prevent the user from receiving real denominations (such as staking rewards).
+
+## Decision
+
+Balances shall be stored per-account & per-denomination under a denomination- and account-unique key, thus enabling O(1) read & write access to the balance of a particular account in a particular denomination.
```

Comment:
O(1) can't be declared in a database backend agnostic method.

Reply:
Hmm, yeah. I was thinking primarily w.r.t. gas costs, which don't incorporate this difference (but perhaps they should, though it seems like it would be complex). At least the complexity is logarithmic, not linear. Regardless, we should clarify this in the ADR.

Reply:
imo gas costs should incorporate this. This isn't an issue in the EVM due to a patricia trie having bounded depth, which is not true of IAVL.